core/src/replay_stage.rs

```diff
@@ -1337,23 +1337,37 @@ impl ReplayStage {
             }
             assert_eq!(*bank_slot, bank.slot());
             if bank.is_complete() {
-                bank_progress.replay_stats.report_stats(
-                    bank.slot(),
-                    bank_progress.replay_progress.num_entries,
-                    bank_progress.replay_progress.num_shreds,
-                );
-                did_complete_bank = true;
-                info!("bank frozen: {}", bank.slot());
-                bank.freeze();
-                heaviest_subtree_fork_choice
-                    .add_new_leaf_slot(bank.slot(), Some(bank.parent_slot()));
-                if let Some(sender) = bank_notification_sender {
-                    sender
-                        .send(BankNotification::Frozen(bank.clone()))
-                        .unwrap_or_else(|err| warn!("bank_notification_sender failed: {:?}", err));
-                }
+                if !blockstore.has_duplicate_shreds_in_slot(bank.slot()) {
```

Comment:
Do you think it's worth it to check this earlier in the replay of this slot here (maybe saves some time replaying this slot?): https://github.com/solana-labs/solana/pull/13992/files#diff-6d8458bb2e53158ac472a9ad4709e6a0a52b75d930c019013298e8acda133828R1286

On the flip side, it incurs the cost of the extra check while replaying, which wouldn't be much if our rocks wrapper actually supported a key check without loading the value: https://github.com/rust-rocksdb/rust-rocksdb/issues/439#issuecomment-740763958

Reply:
I think we can revisit it later, at least now with just checking once it shouldn't be much more expensive in the normal case.

Reply:
@carllin @sakridge 

> On the flip side, it incurs the cost of the extra check while replaying, which might be more efficient once the rocks wrapper supports a key check without loading the value: rust-rocksdb/rust-rocksdb#439 (comment)

This is finally landed in upstream and available in our codebase starting from: #18904

... but it look like the new method incurs the possibility of false positive:

https://docs.rs/rocksdb/0.17.0/rocksdb/struct.DBWithThreadMode.html#method.key_may_exist_cf_opt:

> Returns false if the given key definitely doesn’t exist in the specified column family, otherwise returns true.

Reply:
dang, well that false positive makes that functionality a whole lot less useful :|